### PR TITLE
Routing for using Delegate UIs from identify

### PR DESCRIPTION
**Pull request by maxtaco**
- Make a protocol to register delegate UIs
- Plumb to delegate UI from service/identify
- Clean up IdentifyArg
- Contextify more of libkb for testing
- End-to-end test

Closes CORE-2055

cc: @keybase/react-hackers for a delegated UI in action (see the RPCs in the test `delegate_ui_test.go` for an example.

### CODE REVIEW

File: go/systests/delegate_ui_test.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		case "twitter":
+			d.launchedTwitter = true
+		case "github":
+			d.launchedGithub = true
+		}
+	}
+	return nil
+}
+func (d *delegateUI) DisplayTrackStatement(context.Context, keybase1.DisplayTrackStatementArg) error {
+	if err := d.checkStarted(); err != nil {
```
Comment (MarcoPolo) [inline go/systests/delegate_ui_test.go]: just curious, can you not just return `d.checkStarted()`?

Comment (maxtaco) [inline go/systests/delegate_ui_test.go]: Would be fine!

File: go/systests/delegate_ui_test.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	}
+	if d.finished {
+		return d.setError(fmt.Errorf("Can't run UI after finish() was called"))
+	}
+	return nil
+}
+
+func (d *delegateUI) DelegateIdentifyUI(context.Context) (int, error) {
+	d.delegated = true
+	return 1, nil
```
Comment (chrisnojima) [inline go/systests/delegate_ui_test.go]: returning 1 here is always ok?

Comment (maxtaco) [inline go/systests/delegate_ui_test.go]: It should be for the test, yeah...

File: go/client/cmd_pgp_encrypt.go [hunk 3]
```diff
@@ -15,13 +15,13 @@ import (
 	rpc "github.com/keybase/go-framed-msgpack-rpc"
 )
 
-func NewCmdPGPEncrypt(cl *libcmdline.CommandLine) cli.Command {
+func NewCmdPGPEncrypt(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
```
Comment (chrisnojima) [inline go/client/cmd_pgp_encrypt.go]: is there a better way to do this in go? like actual dependency injection vs having to pass this everywhere?

**Comment by patrickxb** [comment]
Is the contextification of pgp commands necessary for this PR?

**Comment by maxtaco** [comment]
it came about in testing and following it down.

File: go/libkb/globals.go [hunk 4]
```diff
@@ -229,6 +240,10 @@ func (g *GlobalContext) Shutdown() error {
 			g.NotifyRouter.Shutdown()
 		}
 
+		if g.ConnectionManager != nil {
+			g.ConnectionManager.Shutdown()
+		}
+
```
Comment (patrickxb) [inline go/libkb/globals.go]: shut down ui router too

File: go/libkb/notify_router.go [hunk 5]
```diff
@@ -32,14 +32,13 @@ type NotifyRouter struct {
 
 // NewNotifyRouter makes a new notification router; we should only
 // make one of these per process.
-func NewNotifyRouter(g *GlobalContext) *NotifyRouter {
+func NewNotifyRouter(cm *ConnectionManager) *NotifyRouter {
 	ret := &NotifyRouter{
```
Comment (patrickxb) [inline go/libkb/notify_router.go]: what happened to contextified?

File: go/systests/delegate_ui_test.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+	ch        chan error
+	delegated bool
+	started   bool
+	finished  bool
+
+	launchedGithub  bool
+	foundGithub     bool
+	launchedTwitter bool
+	foundTwitter    bool
+}
```
Comment (patrickxb) [inline go/systests/delegate_ui_test.go]: sat what interface this implements

**Comment by patrickxb** [comment]
LGTM after comments

**Pull request merged**